### PR TITLE
MC rate controller: add low-pass filter for D-term

### PR DESCRIPTION
**Pull request by bkueng**
Simpler version of https://github.com/PX4/Firmware/pull/7698.

This adds a butterworth low-pass filter to the D-term of the rate controller.
The idea behind this is: The D-term uses the derivative of the rate and thus is the most susceptible to noise. Therefore, using a D-term filter allows to decrease the driver-level filtering, which leads to reduced control latency and permits to increase the P gains.

### Output Noise
I was using `IMU_GYRO_CUTOFF=90` for my testing. I think I could go even higher, but I should also reduce the on-chip filtering (which I did not do for these tests).

With `IMU_GYRO_CUTOFF=90`, I have visible noise on the outputs, but the quad still flies. I tested different values for the D-term cutoff filter frequency, and it reduces the noise on the output indeed. I found for my quad a cutoff frequency of 70 removes most of the noise.
This is the FFT for actuator outputs without D-term filter (first 3 motors):
fft_actuator_outputs_mc_dterm_cutoff 0

This is using D-term filter frequency = 30:
fft_actuator_outputs_mc_dterm_cutoff 30

### Tuning Gains
I was flying with a P gain of 0.18 before, and was able to increase it to 0.28 now. I tested the same gains against the state of master (`IMU_GYRO_CUTOFF=30`, `MC_DTERM_CUTOFF=0`), and it has strong, visible oscillations, meaning it doesn't fly:
- master: https://logs.px4.io/plot_app?log=980137fc-7bae-48b5-9b93-238989e1ecf4
- this PR with IMU_GYRO_CUTOFF=90, MC_DTERM_CUTOFF=70: https://logs.px4.io/plot_app?log=06ba3019-f8fb-40a8-8fba-5cdc789d07c3

The filter is disabled by default, thus it should not be problematic to merge. We should evaluate the whole filtering pipeline and default values once we have https://github.com/PX4/Firmware/pull/8731.
In future, we can also experiment with different lp filters.

@AndreasAntener @bresch @MaEtUgR @RomanBapst  @Stifael

### CODE REVIEW

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 1]
```diff
@@ -540,6 +559,14 @@ MulticopterAttitudeControl::parameters_update()
 	param_get(_params_handles.pitch_rate_ff, &v);
 	_params.rate_ff(1) = v;
 
+	param_get(_params_handles.d_term_cutoff_freq, &_params.d_term_cutoff_freq);
+	_lp_filters_d[0].set_cutoff_frequency(_loop_update_rate_hz, _params.d_term_cutoff_freq);
+	_lp_filters_d[1].set_cutoff_frequency(_loop_update_rate_hz, _params.d_term_cutoff_freq);
+	_lp_filters_d[2].set_cutoff_frequency(_loop_update_rate_hz, _params.d_term_cutoff_freq);
+	_lp_filters_d[0].reset(_rates_prev(0));
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: Reset on any param change?

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 2]
```diff
@@ -998,13 +1025,20 @@ MulticopterAttitudeControl::control_attitude_rates(float dt)
...
+		_lp_filters_d[2].apply(rates(2)));
+
 	_att_control = rates_p_scaled.emult(rates_err) +
-		       _rates_int +
-		       rates_d_scaled.emult(_rates_prev - rates) / dt +
+		       _rates_int -
+		       rates_d_scaled.emult(rates_filtered - _rates_prev_filtered) / dt +
 		       _params.rate_ff.emult(_rates_sp);
 
 	_rates_sp_prev = _rates_sp;
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: I don't think _rates_sp_prev is used anywhere?

**Comment by dagar** [comment]
Asking out of curiosity, did you consider BlockDerivative?
https://github.com/PX4/Firmware/blob/master/src/lib/controllib/BlockDerivative.hpp

File: src/modules/mc_att_control/mc_att_control_main.cpp [hunk 3]
```diff
@@ -167,11 +168,16 @@ class MulticopterAttitudeControl
 	perf_counter_t	_loop_perf;			/**< loop performance counter */
 	perf_counter_t	_controller_latency_perf;
 
-	math::Vector<3>		_rates_prev;	/**< angular rates on previous step */
-	math::Vector<3>		_rates_sp_prev; /**< previous rates setpoint */
+	math::LowPassFilter2p _lp_filters_d[3];                      /**< low-pass filters for D-term (roll, pitch & yaw) */
+	static constexpr const float initial_update_rate_hz = 250.f; /**< loop update rate used for initialization */
```
Comment (dagar) [inline src/modules/mc_att_control/mc_att_control_main.cpp]: I think we're soon going to need a mechanism to set this system wide (single param or programmatically).

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 1 above)
Comment (bkueng) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: Right, I changed it to update only when the frequency changes.

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 2 above)
Comment (bkueng) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: yep, removed.

File: src/modules/mc_att_control/mc_att_control_main.cpp (hunk 3 above)
Comment (bkueng) [reply src/modules/mc_att_control/mc_att_control_main.cpp]: This is only for initialization. A module needs to be able to determine this on its own and at runtime.

**Comment by bkueng** [comment]
> Asking out of curiosity, did you consider BlockDerivative?

I've seen it before, but forgot about it. It's using a different filter, and comes with the typical Block* overhead. I think it can be useful to have a separate derivative class, but since the derivative is only used in one place, I don't see it being useful in that case.

**Comment by mhkabir** [comment]
I'm seeing an incredible performance increase on my racer after applying this PR and adjusting cutoff frequencies! Thanks @bkueng!

**Comment by bkueng** [comment]
Awesome! Do you have a video & log? :grimacing:

**Comment by hamishwillee** [comment]
@bkueng Is this something we could usefully add docs for in devguide? Perhaps part of a [rate controller diagram](https://dev.px4.io/en/flight_stack/controller_diagrams.html#controller-diagrams)?

**Comment by hamishwillee** [comment]
@bkueng No, not yet (FYI we do have an[ FW attitude controller going in soon](https://github.com/PX4/Devguide/pull/464)). Perhaps worth talking to @bresch about whether you can work together on that?

**Comment by MaEtUgR** [comment]
@bkueng You're the man making it happen! This filtering process is super important, thanks a lot!

**Pull request merged**